File: sdk/python/hitechcloud/resources/domains.py
"""Domains resource (25 endpoints)"""

from . import BaseResource


class DomainsResource(BaseResource):
    """Domains: WHOIS, nameservers, EPP, registration, contact, email forwarding"""

    def list(self, page: int = None, per_page: int = None) -> dict:
        """GET /api/domain - List domains under your account"""
        params = {}
        if page is not None:
            params["page"] = page
        if per_page is not None:
            params["per_page"] = per_page
        return self._http.get("/api/domain", params)

    def get(self, domain_id: int) -> dict:
        """GET /api/domain/{id} - Get domain details"""
        return self._http.get(f"/api/domain/{domain_id}")

    def get_by_name(self, name: str) -> dict:
        """GET /api/domain/name/{name} - Get domain details by name"""
        return self._http.get(f"/api/domain/name/{name}")

    def whois(self, domain: str) -> dict:
        """GET /api/whois/{domain} - WHOIS lookup"""
        return self._http.get(f"/api/whois/{domain}")

    def whois_lookup(self, domain: str) -> dict:
        """GET /api/whoislookup/{domain} - Perform WHOIS lookup"""
        return self._http.get(f"/api/whoislookup/{domain}")

    def check_availability(self, domain: str) -> dict:
        """GET /api/domain/availability/{domain} - Check domain availability"""
        return self._http.get(f"/api/domain/availability/{domain}")

    def get_nameservers(self, domain_id: int) -> dict:
        """GET /api/domain/{id}/ns - Get domain nameservers"""
        return self._http.get(f"/api/domain/{domain_id}/ns")

    def update_nameservers(self, domain_id: int, ns1: str = None, ns2: str = None,
                           ns3: str = None, ns4: str = None) -> dict:
        """PUT /api/domain/{id}/ns - Update domain nameservers"""
        data = {}
        if ns1:
            data["ns1"] = ns1
        if ns2:
            data["ns2"] = ns2
        if ns3:
            data["ns3"] = ns3
        if ns4:
            data["ns4"] = ns4
        return self._http.put(f"/api/domain/{domain_id}/ns", data)

    def get_epp_code(self, domain_id: int) -> dict:
        """GET /api/domain/{id}/epp - Get domain EPP code"""
        return self._http.get(f"/api/domain/{domain_id}/epp")

    def sync(self, domain_id: int) -> dict:
        """GET /api/domain/{id}/sync - Synchronize domain"""
        return self._http.get(f"/api/domain/{domain_id}/sync")

    def get_lock(self, domain_id: int) -> dict:
        """GET /api/domain/{id}/reglock - Get domain lock status"""
        return self._http.get(f"/api/domain/{domain_id}/reglock")

    def update_lock(self, domain_id: int, reglock: bool) -> dict:
        """PUT /api/domain/{id}/reglock - Update domain lock"""
        return self._http.put(f"/api/domain/{domain_id}/reglock", {"reglock": reglock})

    def update_id_protection(self, domain_id: int, idprotection: bool) -> dict:
        """PUT /api/domain/{id}/idprotection - Update ID protection"""
        return self._http.put(f"/api/domain/{domain_id}/idprotection", {
            "idprotection": idprotection,
        })

    def get_contact(self, domain_id: int) -> dict:
        """GET /api/domain/{id}/contact - Get domain contact info"""
        return self._http.get(f"/api/domain/{domain_id}/contact")

    def update_contact(self, domain_id: int, **kwargs) -> dict:
        """PUT /api/domain/{id}/contact - Update domain contact info"""
        return self._http.put(f"/api/domain/{domain_id}/contact", kwargs)

    def get_email_forwarding(self, domain_id: int) -> dict:
        """GET /api/domain/{id}/email/forwarding - Get email forwarding"""
        return self._http.get(f"/api/domain/{domain_id}/email/forwarding")

    def update_email_forwarding(self, domain_id: int, entries: list) -> dict:
        """PUT /api/domain/{id}/email/forwarding - Update email forwarding"""
        return self._http.put(f"/api/domain/{domain_id}/email/forwarding", {
            "entries": entries,
        })

    def get_available_tlds(self) -> dict:
        """GET /api/domain/order - List available TLDs"""
        return self._http.get("/api/domain/order")

    def order(self, domain: str, years: int = 1, **kwargs) -> dict:
        """POST /api/domain/order - Order new domain"""
        data = {"domain": domain, "years": years, **kwargs}
        return self._http.post("/api/domain/order", data)

    def renew(self, domain_id: int, years: int = 1) -> dict:
        """POST /api/domain/{id}/renew - Renew domain"""
        return self._http.post(f"/api/domain/{domain_id}/renew", {"years": years})

    def get_tld_form(self, tld_id: int) -> dict:
        """GET /api/domain/order/{id}/form - Get additional data for TLD"""
        return self._http.get(f"/api/domain/order/{tld_id}/form")

    def get_documents(self, domain_id: int) -> dict:
        """GET /api/domain/{id}/documents - Get domain documents"""
        return self._http.get(f"/api/domain/{domain_id}/documents")

File: sdk/ruby/lib/hitechcloud/resources/domains.rb
module HiTechCloud
  module Resources
    class Domains < Base

        # WHOIS lookup
        def whois(domain)
          @http.get("/api/whois/%{domain}" % { domain: domain })
        end

        # List domains
        def domains
          @http.get("/api/domain")
        end

        # Get domain
        def domain(id)
          @http.get("/api/domain/%{id}" % { id: id })
        end

        # Get domain by name
        def domain_by_name(name)
          @http.get("/api/domain/name/%{name}" % { name: name })
        end

        # Get nameservers
        def nameservers(id)
          @http.get("/api/domain/%{id}/ns" % { id: id })
        end

        # Update nameservers
        def update_nameservers(id, data)
          @http.put("/api/domain/%{id}/ns" % { id: id }, data)
        end

        # Get EPP code
        def epp_code(id)
          @http.get("/api/domain/%{id}/epp" % { id: id })
        end

        # Sync domain
        def sync_domain(id)
          @http.get("/api/domain/%{id}/sync" % { id: id })
        end

        # Get reglock
        def reglock(id)
          @http.get("/api/domain/%{id}/reglock" % { id: id })
        end

        # Update reglock
        def update_reglock(id, data)
          @http.put("/api/domain/%{id}/reglock" % { id: id }, data)
        end

        # Update ID protection
        def id_protection(id, data)
          @http.put("/api/domain/%{id}/idprotection" % { id: id }, data)
        end

        # Get domain contacts
        def domain_contacts(id)
          @http.get("/api/domain/%{id}/contact" % { id: id })
        end

        # Update domain contacts
        def update_contacts(id, data)
          @http.put("/api/domain/%{id}/contact" % { id: id }, data)
        end

        # Get email forwarding
        def email_forwarding(id)
          @http.get("/api/domain/%{id}/email/forwarding" % { id: id })
        end

        # Update email forwarding
        def update_email_forwarding(id, data)
          @http.put("/api/domain/%{id}/email/forwarding" % { id: id }, data)
        end

        # Check availability
        def availability(domain)
          @http.get("/api/domain/availability/%{domain}" % { domain: domain })
        end

        # List domain orders
        def orders
          @http.get("/api/domain/order")
        end

        # Create domain order
        def create_order(data)
          @http.post("/api/domain/order", data)
        end

        # Get order form
        def order_form(id)
          @http.get("/api/domain/order/%{id}/form" % { id: id })
        end

        # Renew domain
        def renew_domain(id, data)
          @http.post("/api/domain/%{id}/renew" % { id: id }, data)
        end

        # WHOIS lookup v2
        def whois_lookup(domain)
          @http.get("/api/whoislookup/%{domain}" % { domain: domain })
        end

        # Get domain documents
        def documents(id)
          @http.get("/api/domain/%{id}/documents" % { id: id })
      end
    end
  end
end
